### PR TITLE
decorating a class

### PR DESCRIPTION
**Pull request by Mekuseo**
- Create a class `Nameable`.
   - Implement a method called `correct_name` that will raise a `NotImplementedError`.

#### Turn your Person class to Nameable
- Make sure that your Person class inherits from Nameable
- Make sure that this class has a method `correct_name` implemented. It should simply return the name attribute.

#### Prepare base Decorator
- Make sure that it inherits from Nameable.
- In the constructor assign a nameable object from params to an instance variable.
- Implement the `correct_name` method that returns the result of the correct_name method of the `@nameable`.

#### Prepare CapitalizeDecorator and TrimmerDecorator
- For the CapitalizeDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that capitalizes the output of `@nameable.correct_name`.
- For the TrimmerDecorator:
    - Create a class that inherits from the base Decorator class.
    - Implement a method `correct_name` that makes sure that the output of `@nameable.correct_name` has a maximum of 10 characters. If it's longer it should trim the word.

### CODE REVIEW

File: rubyClasses/trimmer_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class TrimmerDecorator < Decorator
+  def correct_name
+    name = @nameable.correct_name
+    name.length > 10 ? name[0, 10] : name
```
Comment (BregornOriginal) [inline rubyClasses/trimmer_decorator.rb]: Good work implementing the TrimmerDecorator class

Remember that when we count with arrays, we count from 0

- Kindly modify the conditional to include only the first 10 chars of the word `[0, 9]` 🥇

**Review by BregornOriginal — CHANGES_REQUESTED** [review]
Hi @Mekuseo 

Good job so far!

There are some issues that you still need to work on to go to the next project but you are almost there!

giphy (4)

## Highlights

✅ Linters are passing
✅ Good commit history
✅ Excellent Pull Request description
✅ Correct use of Gitflow
✅ Excellent work implementing nameable 💯 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: rubyClasses/capitalize_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class CapitalizeDecorator < Decorator
+  def correct_name
+    @nameable.correct_name.capitalize
+  end
+end
```
Comment (rotimiazeez) [inline rubyClasses/capitalize_decorator.rb:7]: Please ensure your capitalize_decorator class inherits the Decorator class properly by adding `require_relative 'decorator'` to the top of your class 👍🏽

File: rubyClasses/decorator.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Decorator < Nameable
+  def initialize(nameable)
+    super()
+    @nameable = nameable
+  end
+
+  def correct_name
+    @nameable.correct_name
+  end
+end
```
Comment (rotimiazeez) [inline rubyClasses/decorator.rb:12]: Please ensure your `decorator` class inherits the `nameable` class properly by adding `require_relative 'nameable'` to the top of your class 👍🏽

File: rubyClasses/trimmer_decorator.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class TrimmerDecorator < Decorator
+  def correct_name
+    name = @nameable.correct_name
+    name.length > 10 ? name[0, 9] : name
+  end
+end
```
Comment (rotimiazeez) [inline rubyClasses/trimmer_decorator.rb:8]: Please ensure your `trimmer_decorator` class inherits the `Decorator` class properly by adding `require_relative 'decorator'` to the top of your class 👍🏽

File: rubyClasses/person.rb [hunk 5]
```diff
@@ -1,7 +1,8 @@
-class Person
+class Person < Nameable
```
Comment (rotimiazeez) [inline rubyClasses/person.rb:5]: Your person class should be as shown in the below example:
```
require_relative 'nameable'
require_relative 'capitalize_decorator'
require_relative 'trimmer_decorator'

class Person < Nameable
  attr_accessor :id, :name, :age

  def initialize(age, name = 'Unknown', parent_permission: true)
    super()
    @id = Random.rand(1..1000)
    @name = name
    @age = age
    @parent_permission = parent_permission
  end

  def can_use_services?
    of_age? || @parent_permission
  end

  def correct_name
    name
  end

  private

  def of_age?
    @age >= 18
  end
end

person = Person.new(22, 'maximilianus')
puts person.correct_name
 capitalizedPerson = CapitalizeDecorator.new(person)
puts capitalizedPerson.correct_name
 capitalizedTrimmedPerson = TrimmerDecorator.new(capitalizedPerson)
puts capitalizedTrimmedPerson.correct_name
```

Comment (rotimiazeez) [reply rubyClasses/person.rb:5]: Please try running `ruby rubyClasses/person.rb` before submitting to be sure your class is decorated 😉

**Review by rotimiazeez — CHANGES_REQUESTED** [review]
Hi @Mekuseo 😉, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## Highlights 🔥
- Good Job decorating the `person` class 👏🏽
- Good job implementing the `TrimmerDecorator` and `CapitalizeDecorator` 👍🏽
- Professional-looking PR with a descriptive title and summary 😉
- No linter error ✔️ 

## Suggested changes
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me(@rotimiazeez) in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid, unless it is requested otherwise._**

**Review by mukhammadrizooff — APPROVED** [review]
Hi @Mekuseo, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
- No linter errors ✔️ 
- Correct GitFlow ✔️ 
- Good implementations ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**